### PR TITLE
Allow ViewCtorProp in create_mirror[_view]

### PR DESCRIPTION
**Pull request by masterleinad**
Following-up on #4826, #4844, and #5035.

### CODE REVIEW

**Review by nliber — APPROVED** [review]

File: core/unit_test/tools/TestWithoutInitializing.cpp [hunk 1]
```diff
@@ -75,3 +75,38 @@ TEST(kokkosp, create_mirror_no_init) {
...
+                               Kokkos::DefaultExecutionSpace{}),
+            host_view);
+        auto mirror_device_view = Kokkos::create_mirror_view(
+            Kokkos::view_alloc(Kokkos::HostSpace{},
+                               Kokkos::WithoutInitializing),
+            device_view);
+        auto mirror_host_view = Kokkos::create_mirror_view(
+            Kokkos::view_alloc(Kokkos::HostSpace{}, Kokkos::WithoutInitializing,
+                               Kokkos::DefaultExecutionSpace{}),
+            host_view);
```
Comment (crtrott) [inline core/unit_test/tools/TestWithoutInitializing.cpp:103]: can we test the `view_alloc(WithoutInitializing)` too?

File: containers/src/Kokkos_DynamicView.hpp [hunk 2]
```diff
@@ -685,57 +738,64 @@ inline std::enable_if_t<
...
+template <class Space, class T, class... P, class... ViewCtorArgs>
 std::enable_if_t<
     !Impl::MirrorDynamicViewType<Space, T, P...>::is_same_memspace,
     typename Kokkos::Impl::MirrorDynamicViewType<Space, T, P...>::view_type>
 create_mirror_view(const Space& space,
                    const Kokkos::Experimental::DynamicView<T, P...>& src,
-                   const I&... arg_prop) {
-  return Kokkos::create_mirror(arg_prop..., space, src);
+                   const Impl::ViewCtorProp<ViewCtorArgs...>& arg_prop) {
+  return create_mirror(space, src, arg_prop);
```
Comment (crtrott) [inline containers/src/Kokkos_DynamicView.hpp:762]: did you want to call this here? Or rather add the space to the `arg_prop`?

File: containers/src/Kokkos_DynamicView.hpp [hunk 3]
```diff
@@ -604,59 +604,112 @@ struct MirrorDynamicViewType {
...
-  return Impl::create_mirror(src, wi);
+  return Impl::create_mirror(src, Kokkos::view_alloc(wi));
 }
 
 // Create a mirror in a new space
 template <class Space, class T, class... P>
 inline auto create_mirror(
     const Space& space, const Kokkos::Experimental::DynamicView<T, P...>& src) {
-  return Impl::create_mirror(space, src);
+  return Impl::create_mirror(space, src, Impl::ViewCtorProp<>{});
```
Comment (crtrott) [inline containers/src/Kokkos_DynamicView.hpp:694]: should we just pass `view_alloc(space)` in here?

**Review by crtrott — CHANGES_REQUESTED** [review]
Couple questions but largely good. Follow up: can we get rid of some of the overlaods by making ViewCtorProps implicitly constructible from some of the explicit things.

**Comment by masterleinad** [comment]
> Follow up: can we get rid of some of the overloads by making ViewCtorProps implicitly constructible from some of the explicit things.

No, I don't think that's possible. implicit conversions are not considered during type deduction, see https://en.cppreference.com/w/cpp/language/template_argument_deduction#Implicit_conversions. The only thing we could do is to derive from `ViewCtorProps`, see https://godbolt.org/z/48WcM9foa, but I don't think we should do that. In particular, not for execution spaces and memory spaces.

File: containers/src/Kokkos_DynamicView.hpp (hunk 3 above)
Comment (masterleinad) [reply containers/src/Kokkos_DynamicView.hpp:694]: 3f5c6c43aafa38aac4c9540d47af14ecd712c2ff

File: containers/src/Kokkos_DynamicView.hpp (hunk 2 above)
Comment (masterleinad) [reply containers/src/Kokkos_DynamicView.hpp:762]: 3f5c6c43aafa38aac4c9540d47af14ecd712c2ff

File: core/unit_test/tools/TestWithoutInitializing.cpp (hunk 1 above)
Comment (masterleinad) [reply core/unit_test/tools/TestWithoutInitializing.cpp:103]: Sure, see 8c9aefb8bba6f656cedaafc3232a4fd1974fa638.

**Review by crtrott — APPROVED** [review]

**Pull request merged**